### ISSUE TITLE
[fbpick-coarse] Replace tiled coarse inference with global-anchor inference

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#28]
Background

fbpick-coarse を global-anchor resize mode に移行するため、coarse inference を従来の tiled inference から、gather 全体を固定サイズ 3×256×2048 に変換して一度だけ forward する方式へ置き換える。

従来の coarse inference は、おおむね以下の構成だった。

SEG-Y gather
→ H direction windowing
→ W direction tiled inference
→ tile logits の stitch / average
→ per-trace coarse prediction

この方式は任意の trace 数・sample 数に対応できるが、coarse stage が本来見たい gather 全体の大局的情報を tile 境界で失いやすい。

新しい inference では、以下の構成にする。

SEG-Y full gather
→ gap-aware center anchor selection
→ selected anchor traces: Hc=256
→ time resampling: Wc=2048
→ input: (3, 256, 2048)
→ model forward once per gather
→ anchor-level coarse prediction

この issue では、inference 入力経路と model forward を global-anchor 化する。
anchor prediction を元 trace 数・元 sample 座標に戻して .coarse.npz を保存する処理は、後続 issue で扱う。

Goal

run_fbpick_coarse_infer.py / coarse inference pipeline から tiled-W inference を外し、global-anchor inference path に置き換える。

固定契約:

input mode: global_anchor_resize
input shape: (B, 3, 256, 2048)

channels:
  0: waveform
  1: offset_ch
  2: time_ch

inference 時の anchor selection:

mode: center
gap-aware: true
deterministic: true
Prerequisites

この issue は、以下が完了済みであることを前提にする。

[fbpick-coarse] Add global-anchor 256×2048 config and checkpoint contract
[dataset] Add full-gather sampling support
[fbpick-coarse] Implement gap-aware trace anchor selection
[fbpick-coarse] Implement time resampling and FB coordinate projection
[fbpick-coarse] Replace train/validation datasets with global-anchor datasets

前提として利用できる想定 API:

TraceSubsetSampler.draw_full_gather
select_gap_aware_trace_anchors
resample_waveform_time_axis
build_coarse_time_grid
build_time_channel
project_coarse_indices_to_raw_time

実際の関数名は実装済みコードに合わせる。

Scope
1. Add raw global-anchor inference dataset

対象候補:

packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/build_dataset.py
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/infer.py

raw SEG-Y inference 用に、1 primary gather につき 1 sample を返す dataset を追加する。

想定名:

GlobalAnchorCoarseRawInferDataset

または既存命名規則に合わせる。

dataset の流れ:

1. SEG-Y file info を読む
2. primary gather 全体を選ぶ
3. full gather の trace indices / offsets / headers を取得する
4. gap-aware center anchor selection を行う
5. anchor trace waveform を読み込む
6. waveform を Wc=2048 に time resample する
7. waveform を per-trace standardize する
8. offset_ch と time_ch を作る
9. input tensor と inference metadata を返す

返す input shape:

input: (3, 256, 2048)

raw inference なので target は返さない。

2. Use center anchor selection for inference

inference dataset では、anchor selection を必ず deterministic にする。

anchor_mode = "center"

期待する性質:

- 同じ SEG-Y / config / checkpoint から同じ anchor mapping が得られる
- random generator に依存しない
- gap をまたいだ anchor bin を作らない
- pad row は trace_valid=False で識別できる
3. Preserve inference metadata needed for coordinate restoration

後続 issue で anchor prediction を元 trace / 元 sample 座標へ戻すため、dataset item に metadata を含める。

最低限必要な metadata:

source_file
gather_id / primary key value
raw_trace_indices
anchor_raw_indices
anchor_source_pos
anchor_offsets_m
trace_valid
segment_id
segments
raw_time_len
coarse_time_len
dt_sec
dt_eff_sec
raw_to_coarse_factor
coarse_to_raw_factor
offsets_m for full gather
ffid_values, chno_values など既存 .coarse.npz に必要な header values

DataLoader collate の都合で complex object が扱いづらい場合は、以下の方針にする。

- ndarray / tensor 化できるものは ndarray / tensor にする
- segment 情報は start / stop / n_anchor_rows の配列にする
- debug 用 object は batch_size=1 前提で保持する

raw inference の初期実装では batch_size=1 を前提にしてよい。
ただし model forward 自体は B>=1 に対応できる形にする。

4. Replace tiled-W inference loop

対象候補:

packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/infer.py
cli/run_fbpick_coarse_infer.py

以下の tiled inference 依存を coarse inference path から外す。

InferenceGatherWindowsDataset
collate_pad_w_right
TiledWConfig
iter_infer_loader_tiled_w
tile_w
overlap_w
tiles_per_batch
H-window stitching
W-tile stitching

新しい inference loop は通常の forward にする。

概念例:

for batch in loader:
    x = batch["input"].to(device)  # (B, 3, 256, 2048)

    with torch.no_grad():
        logits = model(x)          # (B, 1, 256, 2048)

期待する logits shape:

logits: (B, 1, 256, 2048)

shape が一致しない場合は即エラーにする。

5. Extract anchor-level predictions from logits

model output から、anchor row ごとの coarse-grid prediction を取得する。

処理例:

logits[b, 0] shape: (256, 2048)

anchor_pick_j:
  argmax over time axis
  shape: (256,)

anchor_pmax:
  softmax over time axis の max probability
  shape: (256,)

期待する output object:

@dataclass(frozen=True)
class GlobalAnchorCoarsePrediction:
    anchor_pick_j: np.ndarray      # shape (256,), coarse sample index
    anchor_pmax: np.ndarray        # shape (256,)
    trace_valid: np.ndarray        # shape (256,)
    metadata: ...

trace_valid=False の pad row は、後続処理で無視できるようにする。

この issue では、anchor_pick_j を元 trace 全体に補間する処理は行わない。
それは後続 issue で実装する。

6. Keep AMP / device / batch config behavior

既存 coarse inference の以下の基本挙動は維持する。

device: auto / cuda / cpu
amp: true / false
batch_size
num_workers
use_tqdm
checkpoint loading
checkpoint metadata validation

ただし、raw global-anchor dataset の metadata が複雑な場合は、初期実装では batch_size=1 のみ許可してよい。

その場合は config validation で明確に落とす。

例:

global-anchor coarse inference currently requires infer.batch_size == 1

将来的に batch inference 対応する場合も、metadata collate の仕様を明確にした上で行う。

7. Update infer config contract

対象候補:

examples/config_infer_fbpick_coarse.yaml
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/config.py

global-anchor inference では、以下の tiled-specific config を使わない。

tile_w
overlap_w
tiles_per_batch
overlap_h

この issue で削除できるなら削除する。
既存 parser との段階的移行が必要なら、まず deprecated / ignored 扱いにして、後続 cleanup issue で削除してもよい。

新しい inference config 例:

coarse:
  input_mode: global_anchor_resize

transform:
  trace_len: 256
  time_len: 2048
  standardize_eps: 1.0e-8

trace_anchor:
  gap_ratio: 5.0
  min_gap_m: null
  infer_mode: center

infer:
  ckpt_path: /path/to/best.pt
  device: auto
  batch_size: 1
  num_workers: 0
  amp: false
  use_tqdm: true
8. Preserve checkpoint metadata validation

global-anchor inference では、checkpoint metadata が以下に一致することを確認する。

coarse_input_mode == "global_anchor_resize"
coarse_trace_len == 256
coarse_time_len == 2048
coarse_in_chans == 3

metadata が欠落している旧 tiled coarse checkpoint は、この inference path では失敗させる。

この validation が前段 issue で実装済みなら、この issue では infer loop から確実に呼ばれていることを確認する。

9. Add temporary adapter boundary for downstream output

この issue では .coarse.npz の最終保存までは扱わないため、次 issue へ渡すための境界を明確にする。

候補:

- inference loop が GlobalAnchorCoarsePrediction の list を返す
- または gather ごとに callback / writer に渡す

後続 issue:

[fbpick-coarse] Restore predictions to original coordinates and preserve .coarse.npz contract

で、以下を実装する。

anchor prediction
→ segment 内補間
→ full trace prediction
→ raw sample coordinate restoration
→ .coarse.npz 保存
Out of scope

この issue では以下は実装しない。

- train / validation dataset の global-anchor 化
- gap-aware trace anchor selection utility 自体の実装
- time resampling utility 自体の実装
- anchor prediction の元 trace 数への補間
- coarse sample index から raw sample index への最終保存処理
- .coarse.npz output contract の完成
- physics / fine stage の変更
- model architecture の変更

この issue は、raw coarse inference 入力経路と model forward を tiled 方式から global-anchor 方式へ置き換えることに集中する。

Acceptance criteria
raw coarse inference dataset が global-anchor input を返す。
inference input shape が (3, 256, 2048) である。
DataLoader batch input shape が (B, 3, 256, 2048) である。
inference dataset は full gather path を使い、H-window crop を行わない。
inference dataset は mode="center" の gap-aware anchor selection を使う。
inference dataset は deterministic である。
coarse inference loop から tiled-W inference 依存が外れている。
iter_infer_loader_tiled_w / TiledWConfig / tile_w / overlap_w / tiles_per_batch が global-anchor path で使われない。
model forward は gather ごとに global input へ一度実行される。
logits shape (B, 1, 256, 2048) が検証される。
anchor-level prediction が取得できる。
anchor_pick_j.shape == (256,)
anchor_pmax.shape == (256,)
pad row は trace_valid=False として後続処理に渡せる。
checkpoint metadata validation が infer path で実行される。
旧 tiled coarse checkpoint は global-anchor inference で明確なエラーになる。
example infer config が global-anchor inference 前提に更新されている。
unit tests / smoke tests が追加または更新されている。
既存 tests が通る。
Suggested tests
1. Raw infer dataset returns fixed global-anchor shape

小さな SEG-Y fixture または mock file info を使う。

期待:

sample["input"].shape == (3, 256, 2048)
2. Raw infer dataset has no target

期待:

"target" not in sample

または既存 convention に合わせて、raw inference では target が None であることを確認する。

3. Inference uses center anchors deterministically

同じ dataset item を複数回読む。

期待:

anchor_source_pos が完全一致する
input が完全一致または数値誤差範囲内で一致する
4. Dataset uses full gather path

TraceSubsetSampler.draw_full_gather を使っていることを確認する。

期待:

local subset crop path が呼ばれない
5. Tiled inference dependencies are not used

global-anchor inference config で infer を実行し、以下が使われないことを確認する。

iter_infer_loader_tiled_w
TiledWConfig
tile_w
overlap_w
tiles_per_batch

実装上は import 削除、または mock / monkeypatch で呼ばれないことを確認する。

6. Model forward input shape

dummy model を使って、forward に渡る tensor shape を検証する。

期待:

x.shape == (B, 3, 256, 2048)
7. Logits shape validation

dummy model が正しい shape を返す場合:

logits.shape == (B, 1, 256, 2048)

成功する。

dummy model が不正 shape を返す場合:

logits.shape == (B, 1, 128, 2048)

明確な ValueError で失敗する。

8. Anchor prediction extraction

time channel をそのまま logits にする dummy model を使う。

期待:

anchor_pick_j == 2047 for valid rows
anchor_pmax.shape == (256,)

pad row の扱いは、trace_valid=False として metadata に残ることを確認する。

9. Checkpoint metadata validation is enforced

metadata 欠落 checkpoint または旧 tiled checkpoint を使う。

期待:

ValueError

エラーメッセージに以下を含める。

expected coarse_input_mode='global_anchor_resize'
10. Batch size behavior is explicit

初期実装で batch_size=1 のみ許可する場合:

infer.batch_size > 1

で明確な config error にする。

将来的に batch 対応する場合は、B>1 でも metadata が正しく batch 化されることを test する。

Implementation notes

推奨する構成:

build_dataset.py

- GlobalAnchorCoarseRawInferDataset
  - full gather sampling
  - center anchor selection
  - time resampling
  - 3ch input construction
  - metadata construction

infer.py

- build raw infer dataset
- DataLoader
- model forward
- logits shape validation
- anchor prediction extraction
- return GlobalAnchorCoarsePrediction objects

削除または global-anchor path から切り離すもの:

InferenceGatherWindowsDataset
collate_pad_w_right
TiledWConfig
iter_infer_loader_tiled_w
tile_w
overlap_w
tiles_per_batch

shape / coordinate の式は inference loop に直接書かない。
time resampling と coordinate projection は既存 utility を呼び出す。
